src/Microsoft.Health.Dicom.Api/Features/ByteCounter/ByteCountingStream.cs

```diff
@@ -0,0 +1,134 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+
+namespace Microsoft.Health.Dicom.Api.Features.ByteCounter
+{
+    /// <summary>
+    /// A stream to wrap an underlying stream and counts the number of bytes passed while operating with it.
+    /// </summary>
+    public class ByteCountingStream : Stream
+    {
+        private readonly Stream _stream;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ByteCountingStream"/> class.
+        /// </summary>
+        /// <param name="stream">An underlying stream.</param>
+        public ByteCountingStream(Stream stream)
+        {
+            EnsureArg.IsNotNull(stream, nameof(stream));
+
+            _stream = stream;
+        }
+
+        /// <inheritdoc />
+        public override bool CanRead => _stream.CanRead;
+
+        /// <inheritdoc />
+        public override bool CanSeek => _stream.CanSeek;
+
+        /// <inheritdoc />
+        public override bool CanWrite => _stream.CanWrite;
+
+        /// <inheritdoc />
+        public override long Length => _stream.Length;
+
+        /// <inheritdoc />
+        public override long Position
+        {
+            get { return _stream.Position; }
+            set { _stream.Position = value; }
+        }
+
+        /// <summary>
+        /// The number of bytes that have been written.
+        /// </summary>
+        public long WrittenByteCount { get; private set; }
```

Comment:
Do we need to worry about thread safety here?

Reply:
Hm yeah seems like we would need to worry about thread safety. I could update it to only be incremented by Interlocked.Add or Interlocked.Increment

Although this code is used in health-paas https://microsofthealth.visualstudio.com/Health/_git/health-paas?path=/src/Common/ServicePlatform/ByteCounting/ByteCountingStream.cs&_a=contents&version=GBmaster
So I'm curious why they didn't add anything like that. Maybe I should follow up with them.

Also moving this to healthcare-shared-components per josiahs suggestion - https://github.com/microsoft/healthcare-shared-components/pull/314

Reply:
Followed up with Eric, who originally wrote this file. He said there was no need for them to keep it thread safe but he sees no reason why we can't. 
I've tested & updated it in the other PR. I'll resolve this thread and we can continue the discussion in the other PR if necessary